### PR TITLE
Add support for new platform (IBM i)

### PR DESCRIPTION
**Pull request by ThePrez**
For reference, see lengthy discussions in #1415 and #1444. I found it eventually much more sensible to have IBM i as a completely new module than sharing a codestream with AIX. 
I'll attach the test output shortly. While there is still plenty of test cases that should be cleaned up, the basic functions seem to work well. This fork has been used by several applications for nearly six months, so there is some assurance of quality. 

This new module relies in an external package, but IBM will distribute the package in RPM form as to ensure that does not pose a problem.

### CODE REVIEW

**Comment by ThePrez** [comment]
[testOutput.txt](https://github.com/giampaolo/psutil/files/3469789/testOutput.txt)

Here is the test output. Like I said, it isn't pretty, but if you agree to merging in the new IBM i module, I'll gladly circle around to clean up the test cases (which I haven't yet figured out how to do), perhaps with the help of @jwoehr.

**Comment by ThePrez** [comment]
@giampaolo , would you be amenable to merging this in once we further clean up these test cases?

**Comment by abmusse** [comment]
I needed to edit requirements.txt to get things running.

```diff
diff --git a/requirements.txt b/requirements.txt
index f8754d3..0dc6b80 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -24,11 +24,11 @@ passlib==1.7.1
 Pillow==5.4.1 ; python_version < '3.7' or sys_platform != 'win32'
 Pillow==6.1.0 ; sys_platform == 'win32' and python_version >= '3.7'
 polib==1.1.0
-psutil==5.6.6
-psycopg2==2.7.7; sys_platform != 'win32' and python_version < '3.8'
+psutil==5.5.1
+psycopg2==2.8.5; sys_platform != 'win32' and python_version < '3.8'
 psycopg2==2.8.3; sys_platform == 'win32' or python_version >= '3.8'
 pydot==1.4.1
-python-ldap==3.1.0; sys_platform != 'win32'
+# python-ldap==3.1.0; sys_platform != 'win32'
 pyparsing==2.2.0
 PyPDF2==1.26.0
 pyserial==3.4
```

**Comment by giampaolo** [comment]
What's the test status?

**Comment by ThePrez** [comment]
The test status is still in pretty sad shape (no work has been done since my posted test results, above). 
If you can verify that you're ok with landing this new platform support, I'll work to clean them up further. 
Thanks!

**Comment by giampaolo** [comment]
Just let me have a look at the output first to have an idea of what's implemented and what is not. =)

**Comment by ThePrez** [comment]
@giampaolo , I've done zero work on this since the posted results on Aug 5th.

File: psutil/_psibmi.py [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+
+
+def net_io_counters(pernic=True):
+    cursor = _conn.cursor()
+    cursor.execute("SELECT LINE_DESCRIPTION FROM QSYS2.NETSTAT_INTERFACE_INFO")
+    counters = {}
+    for row in cursor:
+        counters[row[0]] = (0, 0, 0, 0, 0, 0, 0, 0)  # TODO
+    cursor.close()
+    return counters
```
Comment (giampaolo) [inline psutil/_psibmi.py:273]: For this cursor stuff I suggest to use a context manager. Something like this:

```python

import contextlib


@contextlib.contextmanager
def cursor_ctx():
    cursor = _conn.cursor()
    try:
        yield cursor
    finally:
        cursor.close()
```


...and when you have to use it:

```python
with cursor_ctx() as cursor:
    cursor.execute('...')
```

File: psutil/_psibmi.py [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+
+def boot_time():
+    cursor = _conn.cursor()
+    cursor.execute("""SELECT JOB_ACTIVE_TIME
+                      FROM TABLE(QSYS2.JOB_INFO(JOB_USER_FILTER => 'QSYS')) A
+                      WHERE JOB_NAME = '000000/QSYS/SCPF' """)
+    for row in cursor:
+        ret = row[0].timestamp()
+        cursor.close()
+        return ret
```
Comment (giampaolo) [inline psutil/_psibmi.py:347]: perhaps you should use `row = next(cursor)` and avoid the `for` loop

File: psutil/_psibmi.py [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+
+
+def pids():
+    """Returns a list of PIDs currently running on the system."""
+    return cext.list_pids()
+
+
+def pid_exists(pid):
+    """Check for the existence of a unix pid."""
+    return pid in cext.list_pids()
```
Comment (giampaolo) [inline psutil/_psibmi.py:392]: On UNIX, it's usually enough to do this (faster):

```python
pid_exists = _psposix.pid_exists
```

File: psutil/_psibmi.py [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+    EACCES and EPERM in NoSuchProcess or AccessDenied exceptions.
+    """
+    def wrapper(self, *args, **kwargs):
+        try:
+            return fun(self, *args, **kwargs)
+        except EnvironmentError as err:
+            # support for private module import
+            if (NoSuchProcess is None or AccessDenied is None or
+                    ZombieProcess is None):
+                raise
```
Comment (giampaolo) [inline psutil/_psibmi.py:406]: you can get rid of this stuff

File: psutil/_psibmi.py [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+        return PROC_STATUSES.get(code, '?')
+
+    def open_files(self):
+        # TODO rewrite without using procfiles (stat /proc/pid/fd/* and then
+        # find matching name of the inode)
+        p = subprocess.Popen(["/usr/bin/procfiles", "-n", str(self.pid)],
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = p.communicate()
+        if PY3:
+            stdout, stderr = [x.decode(sys.stdout.encoding)
```
Comment (giampaolo) [inline psutil/_psibmi.py:605]: use `_common.decode()`

File: psutil/_psibmi.py [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+                              for x in (stdout, stderr)]
+        if "no such process" in stderr.lower():
+            raise NoSuchProcess(self.pid, self._name)
+        procfiles = re.findall(r"(\d+): S_IFREG.*\s*.*name:(.*)\n", stdout)
```
Comment (giampaolo) [inline psutil/_psibmi.py:609]: You can probably use `re.compile('...')` and use it as a private argument of this function (faster). 
Grep for "re.compile" to see what I mean.

**Comment by giampaolo** [comment]
Me and Jesse already discussed this privately a while ago, but forgot to update this ticket - doing it now.
Despite I was originally OK-ish pushing this forward, I think it makes more sense to make this work develop and mature in its own fork (**[link to psutil IBM i fork](https://github.com/ThePrez/psutil/tree/ibmi-newmodule)**), essentially for the same reasons I gave in https://github.com/giampaolo/psutil/pull/1690#issuecomment-748222016. Current psutil monthly download stats:

```
$ pypinfo --days 30 psutil system
Served from cache: True
Data processed: 0.00 B
Data billed: 0.00 B
Estimated cost: $0.00

| system_name          | download_count |
| -------------------- | -------------- |
| Linux                |     11,084,549 |
| Windows              |        790,397 |
| Darwin               |        232,315 |
| FreeBSD              |          3,499 |
| OS/390               |             88 |
| AIX                  |             84 |
| SunOS                |             74 |
| OpenBSD              |             39 |
| Total                |     12,111,089 |
```

Given that monthly downloads for AIX and OS/390 (both IBM platforms) are < 100 (0.0007% of total), I expect IBM i numbers would be similar. Going forward, if support for more exotic OSes is started and completed, it may make sense to list them somewhere in mainstream psutil (doc or README).

**Pull request closed without merging**